### DISCUSSION TITLE
Public Key Arithmetic Operations

### DISCUSSION THREAD

**Discussion by albertobsd** [albertobsd/keyhunt#136] [General]
Well public key division or subtraction has some interest because we can make some new public keys in a reduced or lower range.

But how this work?
Lets to think in public keys ( _**P(k)**_ ) as a hidden-number, the hidden number is the private key ( _**k**_ ) we can do simple arithmetic operations over the public key as any other number.

## Subtraction or Addition

To perform this operations both elements in the operation need to be a **public key**.

_**P(9)**_ + _**P(3)**_ = _**P(12)**_

Real values compressed 

```
P(3) = 02f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9
P(9) = 03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe
P(12) = 03d01115d548e7561b15c38f004d734633687cf4419620095bc5b0f47070afe85a
```

Using keymath in [ecctools](https://github.com/albertobsd/ecctools) 

```
./keymath 03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe + 3
Result: 03d01115d548e7561b15c38f004d734633687cf4419620095bc5b0f47070afe85a
```
 Note that the `3` is a number, the program internally transform that value to a public key to make the operations in a Eliptic Cuve way.

With the previous examples you can always back one number to another using the reverse opeations in this case subtraction

_**P(12)**_ - _**P(3)**_ = _**P(9)**_

```
./keymath 03d01115d548e7561b15c38f004d734633687cf4419620095bc5b0f47070afe85a - 3
Result: 03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe
```

## Multiplication or Division

To perform this operations one element in the operation need to be a **public key** and the other one need to be a **Scalar** value

Keep doing this with our previous examples _**P(9)**_  ,  _**P(3)**_

_**P(3)**_ * 3 = _**P(9)**_

_**P(3)**_ * 4 = _**P(12)**_

In the previous 2 examples our scalar values were 3 and 4 respectively.

using keymath to do the same:

```
./keymath 02f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9 x 3
Result: 03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe
```

To do the division is the same we can always back to _**P(3)**_  if we divide the _**P(12)**_  by 4 and the _**P(9)**_  by 4

_**P(12)**_  / 4 = _**P(3)**_ 
_**P(9)**_  / 3 = _**P(3)**_ 

Using keymath to do the same:

```
./keymath 03d01115d548e7561b15c38f004d734633687cf4419620095bc5b0f47070afe85a / 4
Result: 02f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9
```

```
./keymath 03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe / 3
Result: 02f9308a019258c31049344f85f89d5229b531c845836f99b08601f113bce036f9
```

### Divisions are tricky
The main problem with division is that if we have _**P(k)**_ where we don't know what the K value is, **then we don't know what value to use as divisor.**

The easy way is divide by the number `2` but since we don't know what is the **k** value, we don't know if it is **EVEN** or **ODD** so we need to perform 2 divisions.

_**P(k)**_ / 2 = ?
( _**P(k)**_ - _**P(1)**_ ) / 2 = ?

With  _**P(k)**_ and (_**P(k)**_ - _**P(1)**_) we ensure that almost one of those values is EVEN and the division is going to be exact, because inexact divisions also produce valid points in the Curve but in a distinct Bit Range also that range is unknow.

Downgrade 3 bits to the _**P(k)**_

This is other method that don't use subtract 1 and divide by two. to do this we need to do 7 subtraction to the _**P(k)**_ one by one

_**P(k)**_ / 8 = ?
( _**P(k)**_ - _**P(1)**_ ) / 8 = ?
( _**P(k)**_ - _**P(2)**_ ) / 8 = ?
( _**P(k)**_ - _**P(3)**_ ) / 8 = ?
( _**P(k)**_ - _**P(4)**_ ) / 8 = ?
( _**P(k)**_ - _**P(5)**_ ) / 8 = ?
( _**P(k)**_ - _**P(6)**_ ) / 8 = ?
( _**P(k)**_ - _**P(7)**_ ) / 8 = ?

Doing this way, we ensure that almost one the subtracted values is divisible by 8 and the result will be exact, the other 7 values will end in other bit range but one of the results will be 3 bits down of the original publickey.

**Comment by Hurd8x** [comment]
Hello, I think we need a property what you find then you divede big private keys, remember you in your video devide a 800000000000000/2 and get exact 50% from 800000000000000, so I think we need addd to our pubkeys any numpber what bigger the our pubkey and after divide this to 2, so, if we have pubkey 120 we need add to them 140 and after we need devide (140+120)/2 and we get "70,1" bitrange...

**Reply by albertobsd** [reply]
>  if we have pubkey 120 we need add to them 140 and after we need devide (140+120)/2 and we get "70,1" bitrange...

It doesn't work in that way.

if you divide by 2 some public key in X range and the division is  Exact you only reduce it by one bit

**Reply by Hurd8x** [reply]
> > if we have pubkey 120 we need add to them 140 and after we need devide (140+120)/2 and we get "70,1" bitrange...
> 
> It doesn't work in that way.
> 
> if you divide by 2 some public key in X range and the division is Exact you only reduce it by one bit

Why 2 ? only one what  = 140+120.

**Reply by albertobsd** [reply]
oo sorry i think that you are talking about the 120 bit puzzle but i was wrong , you are talking of number 120 and number 140, well 
P(120) + P(140)  = P(260)
P(260) / 2 =  P(130) not P(70)

That is something that I call public key averaging, is another method to move the public key near to the center of some range.

**Reply by Hurd8x** [reply]
> oo sorry i think that you are talking about the 120 bit puzzle but i was wrong , you are talking of number 120 and number 140, well P(120) + P(140) = P(260) P(260) / 2 = P(130) not P(70)
> 
> That is something that I call public key averaging, is another method to move the public key near to the center of some range.

Bro, you mistaken ; add in calc 2^140 + 2^120, result will be smaler then 2^141 !!!

**Reply by albertobsd** [reply]
ooo so it was 2^140 and 2^120, yes you are right, the range may continue in the 140 bit space, or maybe in the 141, that depend of the value of the original of both publickeys, but where you want to go with this example?

**Reply by Hurd8x** [reply]
> ooo so it was 2^140 and 2^120, yes you are right, the range may continue in the 140 bit space, or maybe in the 141, that depend of the value of the original of both publickeys, but where you want to go with this example?

Yes !!! And it will be supper if we go with deviding 2, we down from 142 to 71, like in your examples in video...

**Reply by Hurd8x** [reply]
> > ooo so it was 2^140 and 2^120, yes you are right, the range may continue in the 140 bit space, or maybe in the 141, that depend of the value of the original of both publickeys, but where you want to go with this example?
> 
> Yes !!! And it will be supper if we go with deviding 2, we down from 142 to 71, like in your examples in video...

(2^140+2^120)-2^141= 
-fffff000000000000000000000000000000 !!!!

**Reply by albertobsd** [reply]
Well that is only possible if the value of the public key can be divided by 2 , N number of times in exact way.

But random public keys only can be divided some 2 or 3 times in exact way, but usually we don't know is the public key is Even or Odd as i said in the  first post. that is why in my video I subtract 1 in each round of divisions, to get ensure that almost one of the values can be divided  by 2 in an exact way

**Reply by Hurd8x** [reply]
> Well that is only possible if the value of the public key can be divided by 2 , N number of times in exact way.
> 
> But random public keys only can be divided some 2 or 3 times in exact way, but usually we don't know is the public key is Even or Odd as i said in the first post. that is why in my video I subtract 1 in each round of divisions, to get ensure that almost one of the values can be divided by 2 in an exact way

We can change 03 to 02 from start of pibkick key as nd easy convert our pubkey from odd to even...

**Reply by albertobsd** [reply]
Public keys are a point (X,Y) the 02 or 03 is only for the parity of the Y value, and that value doesn't have any related to the parity of the value of the private Key **_k_**

P(k) = k*G

The parity related to the division is the parity of **_k_**, not of the Y value.

**If we found some way to figure what is the parity of the K value through the publickey we can solve any publickey in seconds, but that method doesn't exist yet.**

**Reply by Hurd8x** [reply]
> Public keys are a point (X,Y) the 02 or 03 is only for the parity of the Y value, and that value doesn't have any related to the parity of the value of the private Key **_k_**
> 
> P(k) = k*G
> 
> The parity related to the division is the parity of **_k_**, not of the Y value

We can subtrat 1G from pubkey ang go to or from even(odd)

**Reply by albertobsd** [reply]
Yes, that is why in the example on the video I subtract 1 to the publickeys....

**Comment by kpot87** [comment]
> Doing this way, we ensure that almost one the subtracted values is divisible by 8 and the result will be exact, the other 7 values will end in other bit range but one of the results will be 3 bits down of the original publickey.

Hi @albertobsd  I don’t understand why when we divide by 8 we just reduced 3 bit and not 8?

**Reply by albertobsd** [reply]
> Hi @albertobsd I don’t understand why when we divide by 8 we just reduced 3 bit and not 8?

It is like to divide by 2 but 3 times, because 2^3 = 8

Check this example:

1024/ 2 = 512
512 / 2 = 256
256 / 2 = 128

is the same of 

1024 / 8 = 128

```
1024 in binary is 10000000000
 128 in binary is    10000000
```

3 bits of difference

**Reply by kpot87** [reply]
And when on lower bit range by bsgs we found privat key, what to do next? I mean how to multiply in right way

**Reply by Hurd8x** [reply]
> And when on lower bit range by bsgs we found privat key, what to do next? I mean how to multiply in right way

Yes - multiply

**Reply by albertobsd** [reply]
To if you found some divided publickey you need to do the operations in reverse way and that depend of what of the 8 publickyes you found, lets to make and example:

P(k) / 8 = A
( P(k) - P(1) ) / 8 = B
( P(k) - P(2) ) / 8 = C
( P(k) - P(3) ) / 8 = D
( P(k) - P(4) ) / 8 = E
( P(k) - P(5) ) / 8 = F
( P(k) - P(6) ) / 8 = G
( P(k) - P(7) ) / 8 = H

Lets to say that you found one of those and you get a Private key **_K'_**. First you multiply that by 8, and if you found A the result of that multiplication is you private **_K_**, but if you found some of the other public keys from B to H you need to ADD the value that you previously subtracted


```
P(k) / 8 = A  --  K = K' * 8
( P(k) - P(1) ) / 8 = B   -- K = (K' * 8) +1
( P(k) - P(2) ) / 8 = C   -- K = (K' * 8) +2
( P(k) - P(3) ) / 8 = D   -- K = (K' * 8) +3
( P(k) - P(4) ) / 8 = E   -- K = (K' * 8) +4
( P(k) - P(5) ) / 8 = F   -- K = (K' * 8) +5
( P(k) - P(6) ) / 8 = G   -- K = (K' * 8) +6
( P(k) - P(7) ) / 8 = H   -- K = (K' * 8) +7
```

**Comment by kpot87** [comment]
@albertobsd and it can be used like I video origin pubkey, than - 1 all devise by 2 to reduce one bit and then repeat again…and again… for example if I would like to reduced 5 bit range…

**Reply by albertobsd** [reply]
Yes the example in the video (subtract one and divide by two) can be repeated until you want but in each step you will to duplicate the you number of publickeys than the previous step.

if you want to reduce 5 bits you will end with 32 publickeys, and only one will be valid in that range.

**Reply by kpot87** [reply]
Ok, but if the key is even we just can divide,  for reduce 5 bit just on 5 or 5 times by two?

**Reply by Hurd8x** [reply]
> Yes the example in the video (subtract one and divide by two) can be repeated until you want but in each step you will to duplicate the you number of publickeys than the previous step.
> 
> if you want to reduce 5 bits you will end with 32 publickeys, and only one will be valid in that range.

Albert, I thin need try add to for ex 140 bit pubkey   80 bit pubkey and result devide by 2 and see we get 70,2 bit or not

**Reply by btclovegalaxy** [reply]
> Yes the example in the video (subtract one and divide by two) can be repeated until you want but in each step you will to duplicate the you number of publickeys than the previous step.
> 
> if you want to reduce 5 bits you will end with 32 publickeys, and only one will be valid in that range.

Hello alebertobsd,

Is there any way to find private key space from public key? Each public key has X and Y value so can we find a relation between X and Y and private key space?

**Comment by kpot87** [comment]
@albertobsd question: if I know that price key is EVEN may I divide by 5 bit at once?

**Reply by albertobsd** [reply]
5 bits is divide it by 32, there are several EVEN values that are not divisible by 32, so there is no way to know if the division is exact or not

**Reply by kpot87** [reply]
But have you make a test of even and odd privatkeys, as I see how often you hit even-even?

**Reply by albertobsd** [reply]
Well that is just probabilistic, numbers ending in 20 decimal are divisible 2
20/ 2 = 10
10/ 2 = 5

But 5 is not divisible by 2 so it break the sequence. You can do small simulations in python with some numbers

**Reply by Hurd8x** [reply]
> Well that is just probabilistic, numbers ending in 20 decimal are divisible 2 20/ 2 = 10 10/ 2 = 5
> 
> But 5 is not divisible by 2 so it break the sequence. You can do small simulations in python with some numbers

Worked for me, but max devide to 4, for go from 120 to 110 need devide 10 steps for 4

Big Thank You !!!

**Comment by kpot87** [comment]
Hi @albertobsd! as i understand we can do any math operation with pubkey? can you add , if its possible  х²  and √

**Reply by kpot87** [reply]
thank you @albertobsd  for your great work

**Reply by albertobsd** [reply]
Publickey square is

P(k^2) = k*  P(k) 

Since we dont know what is the K value we can't get that operation. Some similar is for the square root.

**Comment by kpot87** [comment]
And if we add to last 256 bit +2, we start all 256 bit range again, and it’s possible to come back by /? And which range will be if we multiple 256 bit last key?

**Reply by albertobsd** [reply]
Good questions, yes when we have numbers in the 256 bit range and multiply it by  2, 4 or any other number we are going to end in the same 256 bit range or in some times in a bit lower like 248 or othet... The operation module is responsible for that.
About back to the original number im not sure letme do some simulations.

**Reply by Hurd8x** [reply]
> Good questions, yes when we have numbers in the 256 bit range and multiply it by 2, 4 or any other number we are going to end in the same 256 bit range or in some times in a bit lower like 248 or othet... The operation module is responsible for that. About back to the original number im not sure letme do some simulations.

Albert, message me please to my email, I find a new method what I thant talk in private. Superb speed of finding secp256k1 privkey. My email, caracomongos@gmail.com I waiting your answer.
Regards.

**Reply by Noname400** [reply]
> > Хорошие вопросы, да, когда у нас есть числа в 256-битном диапазоне и умножаются его на 2, 4 или любое другое число, мы собираемся закончить в том же 256-битном диапазоне или в некоторых случаях немного ниже, например, 248 или другое ... За это отвечает операционный модуль. Что касается возврата к исходному номеру, я не уверен, что позвольте мне сделать несколько симуляций.
> 
> Альберт, напишите мне, пожалуйста, на мою электронную почту, я нахожу новый метод, о котором я бы не говорил наедине. Великолепная скорость нахождения приватного ключа secp256k1. Моя электронная почта, [caracomongos@gmail.com](mailto:caracomongos@gmail.com) Жду вашего ответа. С Уважением.

не хотите поделится с сообществом ?
ну или хотя бы со мной ))))

**Reply by Hurd8x** [reply]
> > > Хорошие вопросы, да, когда у нас есть числа в 256-битном диапазоне и умножаются его на 2, 4 или любое другое число, мы собираемся закончить в том же 256-битном диапазоне или в некоторых случаях немного ниже, например, 248 или другое ... За это отвечает операционный модуль. Что касается возврата к исходному номеру, я не уверен, что позвольте мне сделать несколько симуляций.
> > 
> > 
> > Альберт, напишите мне, пожалуйста, на мою электронную почту, я нахожу новый метод, о котором я бы не говорил наедине. Великолепная скорость нахождения приватного ключа secp256k1. Моя электронная почта, [caracomongos@gmail.com](mailto:caracomongos@gmail.com) Жду вашего ответа. С Уважением.
> 
> не хотите поделится с сообществом ? ну или хотя бы со мной ))))

First I thant crack with Albert, if we will be success, and get btc's for us, maybe Albert share method after. Sorry. Albert provide mach information how divide publick keys, you can use this info and make some expiriments. Sorry then pioples like Zielar use 400 cpu servers with 8 TB memory, show to public methods what people like Zielar eating faster then method researcher/developer not good idea.  But ok, I say, what secp256k1 128 bits cracks in minutes Yes !!! I try - worked.

**Reply by albertobsd** [reply]
> Albert, message me please to my email, I find a new method what I thant talk in private. Superb speed of finding secp256k1 privkey. My email, [caracomongos@gmail.com](mailto:caracomongos@gmail.com) I waiting your answer. Regards.

Done!

**Reply by Hurd8x** [reply]
> > Albert, message me please to my email, I find a new method what I thant talk in private. Superb speed of finding secp256k1 privkey. My email, [caracomongos@gmail.com](mailto:caracomongos@gmail.com) I waiting your answer. Regards.
> 
> Done!

I emailed You, Luis. Check your email please. Regards !!! ;)

**Reply by dem10** [reply]
> Albert, message me please to my email, I find a new method what I thant talk in private. Superb speed of finding secp256k1 privkey. My email, [caracomongos@gmail.com](mailto:caracomongos@gmail.com) I waiting your answer. Regards.

I think your idea is not worth anything and you just made it up.

**Reply by Delti42** [reply]
> > Альберт, напишите мне, пожалуйста, на мою электронную почту, я нахожу новый метод, о котором я не буду говорить наедине. Великолепная скорость нахождения приватного ключа secp256k1. Моя электронная почта, [caracomongos@gmail.com](mailto:caracomongos@gmail.com) Жду вашего ответа. С Уважением.
> 
> Выполнено!

Did they write something interesting to you?

**Comment by Noname400** [comment]
wind. Thanks for your work. she is gorgeous!
I want to ask ... how is it possible to determine that we are out of range?
For example:

0000000000000000000000000000000000000000000000000000000000000001 # private key HEX
0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798 # public key

We subtract 1 (- 1) and get:
./keymath 0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798 - 1
Result: 020000000000000000000000000000000000000000000000000000000000000000

And if we subtract 2 (- 2), we get:
./keymath 0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798 - 2
Result: 0379be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798

How to understand that you have jumped to an unknown destination?
is it possible to track this event
and what about division? you can track it in division

**Reply by albertobsd** [reply]
The substract of 1 -1 is the point at infinity (0,0)

The substract of 1 - 2 is (-1)  or ( 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364140 )
0379be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798

But only when you know the K value you can track it, if you are working with an  unknow K is impossible to know the range of the destination.

Regards!

**Comment by Noname400** [comment]
THX!

**Comment by kpot87** [comment]
@albertobsd can you explaine more about division?

**Reply by Hurd8x** [reply]
Ты же знаешь русский язык если я не ошибаюсь, и то что Альберт написал по поводу окончаний публичников на 20, поэкспериментируй, я поэкспериментировал и получилось..

Всего доброго !

**Reply by kpot87** [reply]
Привет, а где он это писал? Чёт не могу понять.

**Reply by Hurd8x** [reply]
> Привет, а где он это писал? Чёт не могу понять.

Я хз, ищи поиск брателло...

**Reply by kpot87** [reply]
Ок

**Comment by Noname400** [comment]
А можно немного поточнен

**Reply by Hurd8x** [reply]
Ну ты попробуй с начала, от чего точнее ? У тебя же свой мозг. ЗЫ. Если у тебя есть 850€ и ты хочешь код который сделает специалист в криптографии код который хакает 256 бит приватники пиши свои контакты... Брат это сложно, долго, геморно, есть аттака которая хакает ,sha256:как чемички... Поэтому Эта тема мне не интересна особо... Чтобы адаптировать аттаку к блокчайн 256 нужен дополнительный код мне не хватает половины что бы заплатить с карты на фирме...

**Reply by Hurd8x** [reply]
> А можно немного поточнен

Я могу тебе пример своих Успешных вычислений показать 500€ в btc, устраивает ?

**Reply by Noname400** [reply]
Не интересно.
